### PR TITLE
[OV20] Reference implementation for NV12toRGB and NV12toBGR operations

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - Reference implementation for NV12toRGB and NV12toBGR operations
 - No preprocessing code, just new operations with tests

### Tests:
- ngraph: visitor + type_prop
- template plugin: reference implementation
- inference-engine: shared tests for plugins
- cpu plugin: compare with ref implementation tests

### Tickets:
 - 62174

### CODE REVIEW

File: ngraph/core/include/ngraph/ops.hpp [hunk 1]
```diff
@@ -170,3 +170,5 @@
 #include "ngraph/op/util/op_types.hpp"
 #include "ngraph/op/variadic_split.hpp"
 #include "ngraph/op/xor.hpp"
+#include "openvino/op/nv12_to_bgr.hpp"
+#include "openvino/op/nv12_to_rgb.hpp"
```
Comment (ilya-lavrenov) [inline ngraph/core/include/ngraph/ops.hpp:174]: I believe we cannot use `ov` directly in ngraph. We probably need to create a ngraph duplicate headers for these operations

CC @ilyachur

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
BTW, can we also tests to compare with old G-API preprocessing? 
It's a kind of compatibility tests:
- create pipeline with new preprocessing and perform inference
- configure `CNNNetwork` to accept `NV12Blob` and performance inference on the same data
- the result should be the same. It will guarantee that users migrated from old preprocessing to new one will have the same experience

Probably, it can be done separatelly.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convert_color_nv12.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+const std::vector<ngraph::element::Type> inTypes = {
+        ngraph::element::u8, ngraph::element::f32
+};
+
+const auto testCase_values = ::testing::Combine(
+    ::testing::ValuesIn(inShapes_nhwc),
+    ::testing::ValuesIn(inTypes),
+    ::testing::Values(true, false),
+    ::testing::Values(true, false),
```
Comment (ilya-lavrenov) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convert_color_nv12.cpp:24]: nit: -> `::testing::Bool()`

File: ngraph/core/include/ngraph/ops.hpp (hunk 1 above)
Comment (ilyachur) [reply ngraph/core/include/ngraph/ops.hpp:174]: We already have a new header:
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/openvino/op/ops.hpp

Comment (ilya-lavrenov) [reply ngraph/core/include/ngraph/ops.hpp:174]: yes, we have, but new operation can be added to old header as well, because both API should work

Comment (ilyachur) [reply ngraph/core/include/ngraph/ops.hpp:174]: Sure

File: ngraph/core/include/openvino/op/nv12_to_bgr.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "openvino/op/util/convert_color_nv12_base.hpp"
+
+namespace ov {
+namespace op {
+namespace v8 {
```
Comment (ilyachur) [inline ngraph/core/include/openvino/op/nv12_to_bgr.hpp:11]: Please add aliases to ngraph namespace

File: ngraph/core/include/ngraph/op/nv12_to_rgb.hpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#pragma once
+
+#include "openvino/op/nv12_to_rgb.hpp"
+
+namespace ngraph {
+namespace op {
+namespace v8 {
+using ov::op::v8::NV12toRGB;
+}  // namespace v8
+using v8::NV12toRGB;
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/nv12_to_rgb.hpp:14]: Please don't add such `vN::...` aliases.

File: ngraph/core/reference/include/ngraph/runtime/reference/convert_color_nv12.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                        size_t stride_uv,
+                        ov::op::util::ConvertColorNV12Base::ColorConversion color_format) {
+    // With C++20 - std::endian can be used at compile time
+    auto little_endian = []() -> int {
+        union {
+            int32_t i;
+            char c[4];
+        } u = {0x00000001};
+        return static_cast<int>(u.c[0]);
+    };
```
Comment (ilya-lavrenov) [inline ngraph/core/reference/include/ngraph/runtime/reference/convert_color_nv12.hpp:32]: why do we need it?

File: ngraph/core/include/ngraph/ops.hpp (hunk 1 above)
Comment (ilya-lavrenov) [reply ngraph/core/include/ngraph/ops.hpp]: fixed

File: ngraph/core/include/openvino/op/nv12_to_bgr.hpp (hunk 3 above)
Comment (ilya-lavrenov) [reply ngraph/core/include/openvino/op/nv12_to_bgr.hpp:11]: fixed

File: ngraph/core/include/ngraph/op/nv12_to_rgb.hpp (hunk 4 above)
Comment (ilya-lavrenov) [reply ngraph/core/include/ngraph/op/nv12_to_rgb.hpp:14]: still here..

File: ngraph/core/reference/include/ngraph/runtime/reference/convert_color_nv12.hpp (hunk 5 above)
Comment (nosovmik) [reply ngraph/core/reference/include/ngraph/runtime/reference/convert_color_nv12.hpp:32]: As per https://docs.microsoft.com/en-us/windows-hardware/drivers/display/4-2-0-video-pixel-formats
> If these samples are addressed as a little-endian WORD type, Cb would be in the least significant bits and Cr would be in the most significant bits with the same total stride as the Y samples
It is possible to determine endianness at compile time, however see,s like not all platforms have macro like __BYTE_ORDER__, __ORDER_LITTLE_ENDIAN__, etc. So I decided to check this at runtime, overhead looks minimal comparing to complexity of other operations

File: ngraph/core/include/ngraph/op/nv12_to_rgb.hpp (hunk 4 above)
Comment (nosovmik) [reply ngraph/core/include/ngraph/op/nv12_to_rgb.hpp]: Fixed

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
@jane-intel @ilyachur @ArtemySkrebkov-intel  please, have a look

File: docs/template_plugin/tests/functional/op_reference/convert_color_nv12.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <gtest/gtest.h>
+
+#include <ngraph/ngraph.hpp>
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/convert_color_nv12.cpp:7]: Can we use new headers?

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/convert_color_nv12.hpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "shared_test_classes/base/layer_test_utils.hpp"
+#include "ngraph_functions/builders.hpp"
+#include "ngraph_functions/utils/ngraph_helpers.hpp"
+#include "openvino/op/nv12_to_rgb.hpp"
+
+namespace LayerTestsDefinitions {
+
+using ConvertColorNV12ParamsTuple = std::tuple<
+        ngraph::Shape,                                 // Input Shape
```
Comment (ilyachur) [inline inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/convert_color_nv12.hpp:19]: Can you use ov::Shape and etc?

File: ngraph/core/include/openvino/op/util/convert_color_nv12_base.hpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    /// \param arg_uv       Node that produces the input tensor for UV plane (NHWC layout). 'H' is half of image height,
+    /// 'W' is half of image width, 'C' dimension equals to 2. Channel 0 represents 'U', channel 1 represents 'V'
+    /// channel
+    ///
+    /// \param format       Conversion format.
+    ConvertColorNV12Base(const Output<Node>& arg_y, const Output<Node>& arg_uv, ColorConversion format);
+
+public:
+    OPENVINO_OP("ConvertColorNV12Base", "util");
+    BWDCMP_RTTI_DECLARATION;
```
Comment (ilyachur) [inline ngraph/core/include/openvino/op/util/convert_color_nv12_base.hpp:65]: We don't need to introduce `BWDCMP_RTTI_DECLARATION` here.

File: ngraph/core/include/openvino/op/nv12_to_rgb.hpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+/// \details Conversion of each pixel from NV12 (YUV) to RGB space is represented by following formulas:
+///        R = 1.164 * (Y - 16) + 1.596 * (V - 128)
+///        G = 1.164 * (Y - 16) - 0.813 * (V - 128) - 0.391 * (U - 128)
+///        B = 1.164 * (Y - 16) + 2.018 * (U - 128)
+///        Then R, G, B values are clipped to range (0, 255)
+///
+class OPENVINO_API NV12toRGB : public util::ConvertColorNV12Base {
+public:
+    OPENVINO_OP("NV12toRGB", "opset8", util::ConvertColorNV12Base);
+    BWDCMP_RTTI_DECLARATION;
```
Comment (ilyachur) [inline ngraph/core/include/openvino/op/nv12_to_rgb.hpp:35]: We don't need this macro if the base class also doesn't define it.

File: ngraph/core/include/openvino/op/util/convert_color_nv12_base.hpp (hunk 8 above)
Comment (nosovmik) [reply ngraph/core/include/openvino/op/util/convert_color_nv12_base.hpp:65]: Could you please clarify this? I see that other 'base' classes in op/util do use this macro so I simply followed to other's behavior. Same question is for 'child' classes like NV12toRGB

**Comment by ilya-lavrenov** [comment]
@ilyachur can we merge as is and fix remarks in next PR with cvtColor preprocessing itself?

File: ngraph/core/include/openvino/op/nv12_to_rgb.hpp (hunk 9 above)
Comment (ilya-lavrenov) [reply ngraph/core/include/openvino/op/nv12_to_rgb.hpp]: fixed

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/convert_color_nv12.hpp (hunk 7 above)
Comment (ilya-lavrenov) [reply inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/convert_color_nv12.hpp]: fixed

File: docs/template_plugin/tests/functional/op_reference/convert_color_nv12.cpp (hunk 6 above)
Comment (ilya-lavrenov) [reply docs/template_plugin/tests/functional/op_reference/convert_color_nv12.cpp]: fixed

File: ngraph/core/src/op/util/convert_color_nv12_base.cpp [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+bool ov::op::util::ConvertColorNV12Base::has_evaluate() const {
+    NGRAPH_OP_SCOPE(v0_ConvertColorNV12Base_has_evaluate);
+
+    return is_type_supported(get_input_element_type(0));
+}
+
+bool ov::op::util::ConvertColorNV12Base::is_type_supported(const ov::element::Type& type) const {
+    return type.is_dynamic() || type.is_real() || type == ov::element::u8;
```
Comment (ilyachur) [inline ngraph/core/src/op/util/convert_color_nv12_base.cpp:230]: Why don't we support integer types?

Comment (nosovmik) [reply ngraph/core/src/op/util/convert_color_nv12_base.cpp:230]: I checked this case with @jane-intel while reviewing 'operation specification'. My original proposal was to support only u8 & fp32, and it was updated to 'u8 or any supported floating point type'. See PR #7595

Comment (nosovmik) [reply ngraph/core/src/op/util/convert_color_nv12_base.cpp:230]: @jane-intel can you please comment on this? User can do 'convert_element_type' if needed, but I'm also ok to update just approved specification and update to 'u8 + any_float + any_number_16bit_or_bigger'

**Review by ilyachur — APPROVED** [review]

File: ngraph/core/src/op/util/convert_color_nv12_base.cpp (hunk 10 above)
Comment (jane-intel) [reply ngraph/core/src/op/util/convert_color_nv12_base.cpp:230]: I was convinced that we could decode NV12 only to the list of types we listed in the spec. but if there is another use case -- then let's all them to spec and here also

File: ngraph/core/src/op/util/convert_color_nv12_base.cpp [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+                                  shape_y[N_DIM].is_dynamic() || shape_uv[N_DIM].is_dynamic() ||
+                                      shape_y[N_DIM].get_length() == shape_uv[N_DIM].get_length(),
+                                  "Y input batch shall be same as UV input batch: Y batch = ",
+                                  shape_y[N_DIM].get_length(),
+                                  " UV batch = ",
+                                  shape_uv[N_DIM].get_length());
+        }
+        // Set shape based on UV shape, if Y are dynamic
+        if (shape_uv.rank().is_static()) {
+            if (out_shape[N_DIM].is_dynamic()) {
```
Comment (jane-intel) [inline ngraph/core/src/op/util/convert_color_nv12_base.cpp:122]: why do we need to check that?

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/util/convert_color_nv12_base.cpp (hunk 11 above)
Comment (nosovmik) [reply ngraph/core/src/op/util/convert_color_nv12_base.cpp:122]: If batch for Y is not defined, but batch for UV is defined - then use batch value from UV

File: ngraph/core/src/op/util/convert_color_nv12_base.cpp (hunk 10 above)
Comment (nosovmik) [reply ngraph/core/src/op/util/convert_color_nv12_base.cpp:230]: Currently specification states "u8 or any supported floating-point type." - here is checked exactly this (+ dynamic type for evaluation on inference stage)

File: ngraph/core/src/op/nv12_to_rgb.cpp [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+ov::op::v8::NV12toRGB::NV12toRGB(const Output<Node>& arg_y, const Output<Node>& arg_uv)
+    : util::ConvertColorNV12Base(arg_y, arg_uv, util::ConvertColorNV12Base::ColorConversion::NV12_TO_RGB) {
+    constructor_validate_and_infer_types();
+}
+
+std::shared_ptr<ov::Node> ov::op::v8::NV12toRGB::clone_with_new_inputs(const OutputVector& new_args) const {
+    NGRAPH_OP_SCOPE(v0_NV12toRGB_clone_with_new_inputs);
+    OPENVINO_ASSERT(new_args.size() == 1 || new_args.size() == 2, "NV12toRGB shall have one or two input nodes");
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/core/src/op/nv12_to_rgb.cpp:21]: I used to treat 'must' and 'shall' like a synonyms when dealing with required behavior :-) Like stated in "https://www.rfc-editor.org/rfc/rfc2119.txt"
> 1. MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.

File: ngraph/core/src/op/util/convert_color_nv12_base.cpp (hunk 11 above)
Comment (jane-intel) [reply ngraph/core/src/op/util/convert_color_nv12_base.cpp:122]: we should use ngraph::Dimension::merge

**Comment by ilya-lavrenov** [comment]
@nosovmik we decided to merge. Let's fix @jane-intel 's remarks in PR with preprocessing step itself.

**Pull request merged**